### PR TITLE
Codemirror comp

### PR DESCRIPTION
**Pull request by ellisonbg**
This converts code mirror to a component.

### CODE REVIEW

**Comment by jasongrout** [comment]
I get an error when compiling with `gulp`: `src/virtualdom/CodeMirrorComponent.ts(28,1): 1110 Type expected.`

**Comment by jasongrout** [comment]
The error is fixed.

File: src/virtualdom/CodeMirrorComponent.ts [hunk 1]
```diff
@@ -0,0 +1,76 @@
...
+import ResizeEvent = events.ResizeEvent;
+
+import Point = geometry.Point;
+import Size = geometry.Size;
+
+
+/**
+ * The class name added to code mirror component classes.
+ */
+var CODE_MIRROR_WIDGET_CLASS = 'p-CodeMirrorComponent';
```
Comment (sccolbert) [inline src/virtualdom/CodeMirrorComponent.ts]: _WIDGET_ -> _COMPONENT_

File: src/virtualdom/CodeMirrorComponent.ts [hunk 2]
```diff
@@ -0,0 +1,76 @@
...
+
+/**
+ * A component which hosts a CodeMirror editor.
+ */
+export
+class CodeMirrorComponent extends BaseComponent<ICodeMirrorData> {
+    
+  static tagName = 'div';
+  
+  static className = CODE_MIRROR_WIDGET_CLASS;
```
Comment (sccolbert) [inline src/virtualdom/CodeMirrorComponent.ts]: `_WIDGET_` -> `_COMPONENT_`

File: src/virtualdom/Component.ts [hunk 3]
```diff
@@ -31,62 +31,18 @@ var EVT_RENDER_REQUEST = { type: 'render-request' };
...
+class Component<T extends IVirtualElementData> extends BaseComponent<T> {
 
   /**
    * Dispose of the resources held by the component.
    */
   dispose(): void {
-    this._m_node = null;
-    this._m_data = null;
+    super.dispose();
     this._m_refs = null;
```
Comment (sccolbert) [inline src/virtualdom/Component.ts]: clear all data before calling `super.dispose()`

File: src/virtualdom/BaseComponent.ts [hunk 4]
```diff
@@ -0,0 +1,114 @@
...
+| Distributed under the terms of the BSD 3-Clause License.
+|
+| The full license is in the file LICENSE, distributed with this software.
+|----------------------------------------------------------------------------*/
+module phosphor.virtualdom {
+
+import IIterable = collections.IIterable;
+
+import ICoreEvent = core.ICoreEvent;
+
```
Comment (sccolbert) [inline src/virtualdom/BaseComponent.ts]: two blank lines

File: src/virtualdom/BaseComponent.ts [hunk 5]
```diff
@@ -0,0 +1,114 @@
...
+var emptyObject: any = Object.freeze(Object.create(null));
+
+/**
+ * A singleton empty array.
+ */
+var emptyArray: any[] = Object.freeze([]);
+
+
+/**
+ * A concrete implementation of IComponent that does not use the virtual DOM to
```
Comment (sccolbert) [inline src/virtualdom/BaseComponent.ts]: Single line summary

File: src/virtualdom/CodeMirrorComponent.ts [hunk 6]
```diff
@@ -0,0 +1,76 @@
...
+
+import Point = geometry.Point;
+import Size = geometry.Size;
+
+
+/**
+ * The class name added to code mirror component classes.
+ */
+var CODE_MIRROR_WIDGET_CLASS = 'p-CodeMirrorComponent';
+
```
Comment (sccolbert) [inline src/virtualdom/CodeMirrorComponent.ts]: two blank lines

File: src/virtualdom/CodeMirrorComponent.ts [hunk 7]
```diff
@@ -0,0 +1,76 @@
...
+export interface ICodeMirrorData {
+  config: CodeMirror.EditorConfiguration;
+}
+
```
Comment (sccolbert) [inline src/virtualdom/CodeMirrorComponent.ts]: two blank lines

File: src/virtualdom/CodeMirrorComponent.ts [hunk 8]
```diff
@@ -0,0 +1,76 @@
...
+   * Create the editor for the widget.
+   *
+   * This can be reimplemented by subclasses which require custom
+   * creation of the editor instance. The default implementation
+   * assumes `CodeMirror` is available in the global scope.
+   */
+  protected createEditor(options?: CodeMirror.EditorConfiguration): CodeMirror.Editor {
+    return CodeMirror(this.node, options);
+  }
+
```
Comment (sccolbert) [inline src/virtualdom/CodeMirrorComponent.ts]: one blank line

File: src/virtualdom/Component.ts [hunk 9]
```diff
@@ -31,62 +31,18 @@ var EVT_RENDER_REQUEST = { type: 'render-request' };
 
 
 /**
- * A concrete implementation of IComponent.
+ * A concrete implementation of IComponent that renders content using a virual
```
Comment (sccolbert) [inline src/virtualdom/Component.ts]: single line summary

File: src/virtualdom/BaseComponent.ts (hunk 4 above)
Comment (ellisonbg) [inline src/virtualdom/BaseComponent.ts]: done

File: src/virtualdom/CodeMirrorComponent.ts (hunk 7 above)
Comment (ellisonbg) [inline src/virtualdom/CodeMirrorComponent.ts]: done

File: src/virtualdom/CodeMirrorComponent.ts (hunk 8 above)
Comment (ellisonbg) [inline src/virtualdom/CodeMirrorComponent.ts]: done

File: gulpfile.js [hunk 10]
```diff
@@ -92,7 +92,9 @@ var tsSources = [
   'virtualdom/factory',
   'virtualdom/dom',
   'virtualdom/renderer',
+  'virtualdom/BaseComponent',
   'virtualdom/Component',
+  'virtualdom/CodeMirrorComponent',
```
Comment (sccolbert) [inline gulpfile.js]: remove `widgets/CodeMirrorWidget`

**Comment by sccolbert** [comment]
Sweet. Thanks!

**Pull request merged**